### PR TITLE
Advanced functionality for bundling (taken from LY)

### PR DESCRIPTION
**Pull request by vmednonogov**
## What does this PR do?

This PRs extends functionality for bundling using the same way if it was made for LY

## How was this PR tested?

Tested locally (currently new code is mostly unreachable)

### CODE REVIEW

File: Code/Framework/AzFramework/AzFramework/Archive/Archive.cpp [hunk 1]
```diff
@@ -1128,6 +1128,48 @@ namespace AZ::IO
         return result;
     }
 
+#if defined(CARBONATED)
+    // returns if a pak has already been opened
+    bool Archive::IsPackOpen(AZStd::string_view szPath)
+    {
+        auto szFullPath = AZ::IO::FileIOBase::GetDirectInstance()->ResolvePath(szPath);
+        if (!szFullPath)
```
Comment (aefimov3) [inline Code/Framework/AzFramework/AzFramework/Archive/Archive.cpp]: The name szFullPath is deceiving, it is not char*, but closer to string. You can use .c_str() with it.
I see that you used the code piece from original o3de, but still szSomething typically means char*.
I believe something neutral like fullPath fits better. Optional.

File: Gems/LmbrCentral/Code/Source/Bundling/BundlingSystemComponent.cpp [hunk 2]
```diff
@@ -114,15 +126,130 @@ namespace LmbrCentral
...
+                    {
+                        AZ_Warning("BundlingSystem", false, "Failed to close bundle %s", szPath);
+                    }
+                }
+
+                m_bundleModeBundles.erase(AZStd::remove(m_bundleModeBundles.begin(), m_bundleModeBundles.end(), bundlePath.c_str()), m_bundleModeBundles.end());
+            }
+            else
+            {
+                return;
```
Comment (aefimov3) [inline Gems/LmbrCentral/Code/Source/Bundling/BundlingSystemComponent.cpp]: Quick return is a better option, like
if (!reload)
{
    return;
}
... reload code...
The logic is more clear, the the major code block indent level is less than if/else case,

File: Gems/LmbrCentral/Code/Source/Bundling/BundlingSystemComponent.cpp [hunk 3]
```diff
@@ -114,15 +126,130 @@ namespace LmbrCentral
...
+                    pakOpened = true;
+                    break;
+                }
+            }
+
+            if (!pakOpened)
+                bundleToLoadList.push_back(thisBundle);
+        }
+
+#if defined(CARBONATED)
```
Comment (aefimov3) [inline Gems/LmbrCentral/Code/Source/Bundling/BundlingSystemComponent.cpp]: This is CARBONATED in CARBONATED, redundant.

File: Gems/LmbrCentral/Code/include/LmbrCentral/Bundling/BundlingSystemComponentBus.h [hunk 4]
```diff
@@ -19,6 +19,18 @@ namespace LmbrCentral
         : public AZ::EBusTraits
     {
     public:
+#if defined(CARBONATED)
+        /**
+         * Loads a specific bundle
+         */
+        virtual void LoadBundle([[maybe_unused]] const char* bundlePath, [[maybe_unused]] bool reload)
```
Comment (aefimov3) [inline Gems/LmbrCentral/Code/include/LmbrCentral/Bundling/BundlingSystemComponentBus.h]: Please use single line:
type function() {}

**Review by aefimov3 — CHANGES_REQUESTED** [review]

File: Code/Framework/AzFramework/AzFramework/Archive/Archive.cpp (hunk 1 above)
Comment (vmednonogov) [reply Code/Framework/AzFramework/AzFramework/Archive/Archive.cpp]: Updated

File: Gems/LmbrCentral/Code/Source/Bundling/BundlingSystemComponent.cpp (hunk 2 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Bundling/BundlingSystemComponent.cpp]: Updated

File: Gems/LmbrCentral/Code/Source/Bundling/BundlingSystemComponent.cpp (hunk 3 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Bundling/BundlingSystemComponent.cpp]: Updated

File: Gems/LmbrCentral/Code/include/LmbrCentral/Bundling/BundlingSystemComponentBus.h (hunk 4 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/include/LmbrCentral/Bundling/BundlingSystemComponentBus.h]: Updated

**Review by Twolewis — APPROVED** [review]

**Review by carb-lloyd — APPROVED** [review]

**Review by aefimov3 — APPROVED** [review]

File: Code/Framework/AzFramework/AzFramework/Archive/Archive.cpp [hunk 5]
```diff
@@ -1128,6 +1128,48 @@ namespace AZ::IO
...
     }
 
+#if defined(CARBONATED)
+    // returns if a pak has already been opened
+    bool Archive::IsPackOpen(AZStd::string_view szPath)
+    {
+        auto fullPath = AZ::IO::FileIOBase::GetDirectInstance()->ResolvePath(szPath);
+        if (!fullPath)
+        {
+            AZ_Assert(false, "Unable to resolve path for filepath %.*s", aznumeric_cast<int>(szPath.size()), szPath.data());
```
Comment (alexeykostin) [inline Code/Framework/AzFramework/AzFramework/Archive/Archive.cpp]: You can use  AZ_STRING_ARG(szPath) to simplify.

File: Code/Framework/AzFramework/AzFramework/Archive/Archive.h [hunk 6]
```diff
@@ -190,6 +190,11 @@ namespace AZ::IO
         //returns if a archive exists matching the wildcard
         bool FindPacks(AZStd::string_view pWildcardIn) override;
 
+#if defined(CARBONATED)
+        // returns if a pak has already been opened
+        bool IsPackOpen(AZStd::string_view pName) override;
```
Comment (alexeykostin) [inline Code/Framework/AzFramework/AzFramework/Archive/Archive.h:195]: Quite minor, but declared name is szPath. Better to be consistent.

File: Code/Framework/AzFramework/AzFramework/Archive/IArchive.h [hunk 7]
```diff
@@ -110,6 +110,10 @@ namespace AZ::IO
         virtual bool ClosePacks(AZStd::string_view pWildcard) = 0;
         //returns if a archive exists matching the wildcard
         virtual bool FindPacks(AZStd::string_view pWildcardIn) = 0;
+#if defined(CARBONATED)
+        // returns if a pak has already been opened
+        virtual bool IsPackOpen(AZStd::string_view pName) = 0;
```
Comment (alexeykostin) [inline Code/Framework/AzFramework/AzFramework/Archive/IArchive.h:115]: Quite minor, but declared name is szPath. Better to be consistent.

Also in one more occurrence.

File: Code/Framework/AzFramework/AzFramework/Archive/Archive.cpp (hunk 5 above)
Comment (vmednonogov) [reply Code/Framework/AzFramework/AzFramework/Archive/Archive.cpp]: Replaced to AZ_STRING_ARG

File: Code/Framework/AzFramework/AzFramework/Archive/Archive.h (hunk 6 above)
Comment (vmednonogov) [reply Code/Framework/AzFramework/AzFramework/Archive/Archive.h:195]: Kept here
Renamed in the method definition

File: Code/Framework/AzFramework/AzFramework/Archive/IArchive.h (hunk 7 above)
Comment (vmednonogov) [reply Code/Framework/AzFramework/AzFramework/Archive/IArchive.h:115]: Kept here
Renamed in the method definition

File: Gems/LmbrCentral/Code/Source/Bundling/BundlingSystemComponent.cpp [hunk 8]
```diff
@@ -13,6 +13,7 @@
 #include <AzCore/Interface/Interface.h>
 #include <AzCore/Serialization/Utils.h>
 #include <AzCore/std/string/string_view.h>
+#include <AzCore/std/sort.h>
```
Comment (alexeykostin) [inline Gems/LmbrCentral/Code/Source/Bundling/BundlingSystemComponent.cpp]: It would be better to add `#if defined(CARBONATED)` here

**Review by alexeykostin — APPROVED** [review]
Minor comment

File: Gems/LmbrCentral/Code/Source/Bundling/BundlingSystemComponent.cpp (hunk 8 above)
Comment (vmednonogov) [reply Gems/LmbrCentral/Code/Source/Bundling/BundlingSystemComponent.cpp]: Updated

**Pull request merged**